### PR TITLE
Integrate CNI firewall plugin code

### PR DESCRIPTION
**Pull request by mheon**
Replace our old manual iptables hangling with (modified) vendored code from https://github.com/containernetworking/plugins/pull/75. This should greatly improve our interaction with the system firewall.

### CODE REVIEW

**Comment by mheon** [comment]
@dcbw PTAL - had to modify your code a bit for rootless support and non-Linux builds, but it's basically identical if a bit rearranged.

@baude @rhatdan PTAL

**Comment by mheon** [comment]
bot, retest this please

File: libpod/networking_linux.go [hunk 1]
```diff
@@ -63,14 +63,14 @@ func (r *Runtime) configureNetNS(ctr *Container, ctrNS ns.NetNS) (err error) {
...
-				iptablesDNS("-I", ip.Address.IP.String())
-			}
+	// Add firewall rules to ensure the container has network access.
+	// Will not be necessary once CNI firewall plugin merges upstream.
+	for _, netStatus := range ctr.state.NetworkStatus {
+		firewallConf := &firewall.FirewallNetConf{
+			PrevResult: netStatus,
+		}
+		if err := r.firewallBackend.Add(firewallConf); err != nil {
+			return errors.Wrapf(err, "error adding firewall rules for container %s", ctr.ID())
```
Comment (TomSweeneyRedHat) [inline libpod/networking_linux.go]: Is there an active PR upstream?  If so, might be nice to add a link,if not, ignore this comment.

File: libpod/networking_linux.go [hunk 2]
```diff
@@ -213,15 +201,14 @@ func (r *Runtime) teardownNetNS(ctr *Container) error {
...
-				iptablesDNS("-D", ip.Address.IP.String())
-			}
+	// Remove firewall rules we added on configuring the container.
+	// Will not be necessary once CNI firewall plugin merges upstream.
+	for _, netStatus := range ctr.state.NetworkStatus {
+		firewallConf := &firewall.FirewallNetConf{
+			PrevResult: netStatus,
+		}
+		if err := r.firewallBackend.Del(firewallConf); err != nil {
+			return errors.Wrapf(err, "error removing firewall rules for container %s", ctr.ID())
```
Comment (rhatdan) [inline libpod/networking_linux.go]: Should we continue to attempt to delete other file wall rules and wrap all of the errors together before returning?

**Comment by rhatdan** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
Something strange is happening in the test suites for a while, where /dev/shm seems to be switching?

```
/go/src/github.com/containers/libpod/test/e2e/run_ns_test.go:11
  podman run ipcns test [It]
  /go/src/github.com/containers/libpod/test/e2e/run_ns_test.go:50

  Expected
      <string>: 294298 /dev/shm
  to equal
      <string>: 158124 /dev/shm
```

**Comment by rhatdan** [comment]
bot, retest this please

File: libpod/networking_linux.go (hunk 2 above)
Comment (mheon) [reply libpod/networking_linux.go]: Probably more trouble than it's worth - 99% of our deployments will likely only be using one network here

Comment (rhatdan) [reply libpod/networking_linux.go]: ok.

File: libpod/networking_linux.go (hunk 1 above)
Comment (ashcrow) [reply libpod/networking_linux.go]: https://github.com/containernetworking/plugins/pull/75

**Comment by mheon** [comment]
An update here: this does not appear to resolve our port forwarding issues. Investigating now.

**Comment by mheon** [comment]
Alright, documenting my findings here before I'm off for lunch:
- Remote port forwarding always works, but you may need a system restart to clear the firewall and make it start working
- Local port forwarding works until anything else starts messing with iptables. And I mean anything. Even listing rules via `iptables -nvL` will cause local forwarding to break.

Further details:
After an `iptables -nvL` or `iptables-save`, packets from localhost (to either 127.0.0.1 or any external IP) successfully make it to the CNI-HOSTPORT-DNAT prerouting chain, but either are not forwarded to or are not matched by the per-container CNI-DN-<ID> chain.

**Comment by baude** [comment]
bot, retest this please

**Comment by baude** [comment]
LGTM as is ... I would like to see this merged.  Functionally it is doing what we expected even though more debug on other uses cases and environments is needed.  Merging will also allow us and others to use a COPR package for testing.

**Comment by baude** [comment]
@rh-atomic-bot r+

**Comment by sergeiwaigant** [comment]
Can someone give me an info if this is fixed in 1.4.4 version of podman?
I am facing this issue on my RHEL 7.7 machine at my customers on-prem systems.

**Comment by mheon** [comment]
Definitely should be. This was resolved before we went 1.0.

**Comment by sergeiwaigant** [comment]
I can say that on a fresh RHEL 7.7 system with latest Podman version provided by RedHat I can reproduce this issue.

Steps to reproduce:

- Download latest RHEL 7.7 ISO
- Install as standard server, with or without GUI doesnt matter
- Install Podman
- Run following commands
- Between the both `curl`'s run `firewall-cmd --reload` on another terminal

```
[root@rhel-local sergeiwaigant]# cat /etc/redhat-release
Red Hat Enterprise Linux Server release 7.7 (Maipo)
[root@rhel-local sergeiwaigant]# uname -mrs
Linux 3.10.0-1062.el7.x86_64 x86_64
[root@rhel-local sergeiwaigant]# podman --version
podman version 1.4.4

[root@rhel-local sergeiwaigant]# podman run --rm -ti registry.access.redhat.com/rhel7
[root@b860133fee69 /]# curl ifconfig.co
4x.1xx.2xx.1xx
[root@b860133fee69 /]# curl ifconfig.co
curl: (6) Could not resolve host: ifconfig.co; Unknown error
root@b860133fee69 /]# exit
ERRO[0157] Error deleting network: running [/sbin/iptables -t nat -D POSTROUTING -s 10.88.0.24 -j CNI-f50e48a52fa26aaaad634e9a -m comment --comment name: "podman" id: "b860133fee6983ea2b8bab495e1beffac0f1fd2232db8d923d57c12d67b08649" --wait]: exit status 2: iptables v1.4.21: Couldn't load target `CNI-f50e48a52fa26aaaad634e9a':No such file or directory

Try `iptables -h' or 'iptables --help' for more information.
ERRO[0157] Error while removing pod from CNI network "podman": running [/sbin/iptables -t nat -D POSTROUTING -s 10.88.0.24 -j CNI-f50e48a52fa26aaaad634e9a -m comment --comment name: "podman" id: "b860133fee6983ea2b8bab495e1beffac0f1fd2232db8d923d57c12d67b08649" --wait]: exit status 2: iptables v1.4.21: Couldn't load target `CNI-f50e48a52fa26aaaad634e9a':No such file or directory

Try `iptables -h' or 'iptables --help' for more information.
ERRO[0157] unable to cleanup network for container b860133fee6983ea2b8bab495e1beffac0f1fd2232db8d923d57c12d67b08649: "error tearing down CNI namespace configuration for container b860133fee6983ea2b8bab495e1beffac0f1fd2232db8d923d57c12d67b08649: running [/sbin/iptables -t nat -D POSTROUTING -s 10.88.0.24 -j CNI-f50e48a52fa26aaaad634e9a -m comment --comment name: \"podman\" id: \"b860133fee6983ea2b8bab495e1beffac0f1fd2232db8d923d57c12d67b08649\" --wait]: exit status 2: iptables v1.4.21: Couldn't load target `CNI-f50e48a52fa26aaaad634e9a':No such file or directory\n\nTry `iptables -h' or 'iptables --help' for more information.\n"
```

Regards
-Sergei

**Comment by mheon** [comment]
Firewall reload wiping our rules is a known issue, with the recommended solution being to restart the containers. I recommend opening a bugzilla if this is considered a critical or blocking issue so we can prioritize appropriately.

**Comment by sergeiwaigant** [comment]
Hm its kind of blocking at my client, with a workaround available, though... just fall back to Docker as container runtime for time being. Unless they stick to 7.7 and not go to 8 the Docker package is available.
Then it becomes blocking and crticial without a workaround :D

Is this then now more firewall related or podman / CNI related? Didnt get the underlying problem exactly.
Is there a an open issue on GitHub? couldnt find it...

I will file a bugzilla for this, pointing to the problem.

**Comment by Klaas-** [comment]
@sergeiwaigant did you file a bz? I could not locate one

**Comment by sergeiwaigant** [comment]
@Klaas- nope, not so far... will retry this on a fresh system here once and then file a bz.

**Comment by mdaguete** [comment]
The same behaviour explained by @sergeiwaigant is happening with RH8.1

**Pull request closed without merging**